File: scripts/discovery/scan-all-hooks.js
#!/usr/bin/env node
/**
 * Scan all projects for .claude/settings*.json with hooks.
 *
 * Scans predefined directories for projects containing hooks configuration
 * in settings.json or settings.local.json files.
 *
 * Pattern from scan-all-hooks.py
 */

const fs = require('fs');
const path = require('path');
const { writeJson } = require('../lib/file-utils');

// Directories to scan
const SCAN_DIRS = [
    'C:\\Users\\julien\\OneDrive\\Coding\\_Projets de code',
    'C:\\Users\\julien\\OneDrive\\Coding\\_référentiels de code'
];

// Projects to exclude
const EXCLUDE = ['2025.11 Claude Code MarketPlace'];

/**
 * Find all hooks in settings files within a directory tree.
 *
 * @param {string} basePath - Base directory to scan
 * @returns {Object} Results by project name
 */
function findHooks(basePath) {
    const results = {};

    // Check if base path exists
    if (!fs.existsSync(basePath)) {
        return results;
    }

    const items = fs.readdirSync(basePath);

    for (const item of items) {
        const itemPath = path.join(basePath, item);

        // Skip non-directories
        if (!fs.statSync(itemPath).isDirectory()) {
            continue;
        }

        // Skip excluded directories
        if (EXCLUDE.includes(item)) {
            continue;
        }

        // Skip hidden directories
        if (item.startsWith('.')) {
            continue;
        }

        // Check for .claude directory
        const claudeDir = path.join(itemPath, '.claude');
        if (!fs.existsSync(claudeDir)) {
            continue;
        }

        // Check both settings files
        const settingsFiles = ['settings.json', 'settings.local.json'];

        for (const settingsFile of settingsFiles) {
            const settingsPath = path.join(claudeDir, settingsFile);

            if (fs.existsSync(settingsPath)) {
                try {
                    const content = fs.readFileSync(settingsPath, 'utf8');
                    const data = JSON.parse(content);

                    if (data.hooks && Object.keys(data.hooks).length > 0) {
                        const hooks = data.hooks;
                        const hookSummary = {};

                        // Count hooks per event
                        for (const [event, config] of Object.entries(hooks)) {
                            if (config) {
                                // If it's an array, count items, otherwise count as 1
                                hookSummary[event] = Array.isArray(config) ? config.length : 1;
                            }
                        }

                        if (Object.keys(hookSummary).length > 0) {
                            // Initialize project entry if needed
                            if (!results[item]) {
                                results[item] = {
                                    path: itemPath,
                                    settings_files: []
                                };
                            }

                            results[item].settings_files.push({
                                file: settingsFile,
                                hooks: hookSummary,
                                full_hooks: hooks
                            });
                        }
                    }
                } catch (error) {
                    console.log(`Error reading ${settingsPath}: ${error.message}`);
                }
            }
        }
    }

    return results;
}

/**
 * Scan all configured directories for hooks.
 *
 * @param {string[]} scanDirs - Directories to scan
 * @returns {Object} Results by scan directory
 */
function scanAllDirectories(scanDirs) {
    const allResults = {};

    for (const scanDir of scanDirs) {
        if (fs.existsSync(scanDir)) {
            allResults[scanDir] = findHooks(scanDir);
        } else {
            allResults[scanDir] = {};
        }
    }

    return allResults;
}

/**
 * Main execution function.
 */
async function main() {
    const allResults = scanAllDirectories(SCAN_DIRS);
    let totalProjectsWithHooks = 0;

    // Count total projects with hooks
    for (const scanDir in allResults) {
        totalProjectsWithHooks += Object.keys(allResults[scanDir]).length;
    }

    // Print summary
    console.log('=== HOOKS SCAN RESULTS ===\n');
    console.log(`Total projects with hooks: ${totalProjectsWithHooks}\n`);

    for (const scanDir in allResults) {
        const projects = allResults[scanDir];
        if (Object.keys(projects).length > 0) {
            console.log(`\n--- ${scanDir} ---\n`);

            // Sort projects by name
            const sortedProjects = Object.entries(projects).sort((a, b) => a[0].localeCompare(b[0]));

            for (const [projectName, data] of sortedProjects) {
                console.log(`\n📁 ${projectName}`);

                for (const sf of data.settings_files) {
                    console.log(`   File: ${sf.file}`);

                    // Print hook counts
                    for (const [event, count] of Object.entries(sf.hooks)) {
                        console.log(`      - ${event}: ${count} hook(s)`);
                    }

                    // Print actual hook commands
                    console.log('   Commands:');
                    for (const [event, hooksList] of Object.entries(sf.full_hooks)) {
                        if (Array.isArray(hooksList)) {
                            for (const hookGroup of hooksList) {
                                if (hookGroup && typeof hookGroup === 'object' && hookGroup.hooks) {
                                    for (const h of hookGroup.hooks) {
                                        const cmd = (h.command || '').slice(0, 80);
                                        console.log(`      [${event}] ${cmd}...`);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    // Save JSON
    const outputPath = path.join(__dirname, '..', 'docs', 'hooks-scan-results.json');
    await writeJson(outputPath, {
        total_projects: totalProjectsWithHooks,
        results: allResults
    });
    console.log(`\n\nJSON saved to: ${outputPath}`);
}

// CLI wrapper
/* istanbul ignore next */
if (require.main === module) {
    main().catch(error => {
        console.error('Error:', error.message);
        process.exit(1);
    });
}

module.exports = { findHooks, scanAllDirectories, main };

File: backups/python-legacy-20260112/scripts/discovery/scan-all-hooks.py
#!/usr/bin/env python3
"""Scan all projects for .claude/settings*.json with hooks."""
import os
import json
from pathlib import Path

SCAN_DIRS = [
    Path(r"C:\Users\julien\OneDrive\Coding\_Projets de code"),
    Path(r"C:\Users\julien\OneDrive\Coding\_référentiels de code"),
]
EXCLUDE = ["2025.11 Claude Code MarketPlace"]

def find_hooks(base_path: Path) -> dict:
    """Find all hooks in settings files."""
    results = {}

    for item in base_path.iterdir():
        if not item.is_dir():
            continue
        if item.name in EXCLUDE:
            continue
        if item.name.startswith('.'):
            continue

        claude_dir = item / ".claude"
        if not claude_dir.exists():
            continue

        for settings_file in ["settings.json", "settings.local.json"]:
            settings_path = claude_dir / settings_file
            if settings_path.exists():
                try:
                    with open(settings_path, "r", encoding="utf-8") as f:
                        data = json.load(f)

                    if "hooks" in data and data["hooks"]:
                        hooks = data["hooks"]
                        hook_summary = {}
                        for event, config in hooks.items():
                            if config:
                                hook_summary[event] = len(config) if isinstance(config, list) else 1

                        if hook_summary:
                            if item.name not in results:
                                results[item.name] = {
                                    "path": str(item),
                                    "settings_files": []
                                }
                            results[item.name]["settings_files"].append({
                                "file": settings_file,
                                "hooks": hook_summary,
                                "full_hooks": hooks
                            })
                except Exception as e:
                    print(f"Error reading {settings_path}: {e}")

    return results

def main():
    all_results = {}
    total_projects_with_hooks = 0

    for scan_dir in SCAN_DIRS:
        if scan_dir.exists():
            results = find_hooks(scan_dir)
            all_results[str(scan_dir)] = results
            total_projects_with_hooks += len(results)

    # Print summary
    print(f"=== HOOKS SCAN RESULTS ===\n")
    print(f"Total projects with hooks: {total_projects_with_hooks}\n")

    for scan_dir, projects in all_results.items():
        if projects:
            print(f"\n--- {scan_dir} ---\n")
            for project, data in sorted(projects.items()):
                print(f"\n📁 {project}")
                for sf in data["settings_files"]:
                    print(f"   File: {sf['file']}")
                    for event, count in sf["hooks"].items():
                        print(f"      - {event}: {count} hook(s)")
                    # Show actual hook commands
                    print("   Commands:")
                    for event, hooks_list in sf["full_hooks"].items():
                        if isinstance(hooks_list, list):
                            for hook_group in hooks_list:
                                if isinstance(hook_group, dict) and "hooks" in hook_group:
                                    for h in hook_group["hooks"]:
                                        cmd = h.get("command", "")[:80]
                                        print(f"      [{event}] {cmd}...")

    # Save JSON
    output_path = Path(__file__).parent.parent / "docs" / "hooks-scan-results.json"
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump({
            "total_projects": total_projects_with_hooks,
            "results": all_results
        }, f, indent=2, ensure_ascii=False)
    print(f"\n\nJSON saved to: {output_path}")

if __name__ == "__main__":
    main()
